### ISSUE TITLE
Lightbox: Keyboard trap with screen reader shortcuts 

### ISSUE THREAD

**Issue by RubyDo** [wet-boew/wet-boew#6619]
I have a questions and accessibility concern for the lightbox: 
1- Most of the examples use lightbox to display images and static content. Any accessibility concerns about the use of lightbox to display dynamic forms (in which part of the form change upon user selection) and dynamic datatable (that use the DataTables plugin)? 

2- Keyboard trap with screen reader shortcuts 
Here is the scenario of what happened when a screen reader user: 
- User clicks on a register button that opened a lightbox with a registration form. 
- In the lightbox, user uses Jaws shortcuts to  jump to canada.ca search (to allocate a different information) 
- User goes to form mode, types keyword but no submission functionality 
- User tries to navigate through the main menu and other links (nothing worked)
- User is in a keyboard trap and forced to restart browser

My analysis: 
On a lightbox , a sighted user is aware of the boundaries of  lightbox content, the cancel button and can not access the main page.  
On the other hand, a screen reader does not get all of this information and can access the elements on the main page using screen reader shortcuts.

Suggestion: 
May be the user experience could change if in the examples on the lightbox page http://wet-boew.github.io/wet-boew/demos/lightbox/lightbox-en.html?wbdisable=false include awareness (about being in a lightbox popup, what are the lightbox content boundary and the necessity to click cancel/close to close the popup)
Here is my suggestion:
•	Button  of lightbox (Open in same window popup. Use close modal popup button to close the popup. No screen reader shortcut for Navigation is allowed.)
•	Lightbox popup page structure (Attached a screenshot with the suggestion)
Hidden paragraph - This is the start of the popup page 
button-Close modal popup 
Title h2 (popup content)
Hidden paragraph - End of popup page
Button-Close modal popup 
![lightbox](https://cloud.githubusercontent.com/assets/1933074/6642068/4fa9a214-c978-11e4-810b-e5c63c967cbf.png)

**Comment by RubyDo** [comment]
@pjackson28  I am not sure to whom should I be addressing my issue.

**Comment by LaurentGoderre** [comment]
I don't think that is necessary. The plugin probably need to change to that if the focus leaves the lightbox, it closes it. I don't think this can be done via keyboard alone but using a screen reader can certainly do that.

**Comment by RubyDo** [comment]
@LaurentGoderre  Giving user awareness in the plugin on lighbox open and close actions. 
This is in case the user filled a part of a form on a light box popup, then navigated with screen reader on main page elements then wanting to return to lightbox popup again. (Closing the lightbox without user awareness means losing input filled by the user and may cause confusion.)

**Comment by LaurentGoderre** [comment]
Then perhaps a lightbox is not the best method to do this kind of task. I doubt the proposed solution will do much to reduce the confusion.

**Comment by RubyDo** [comment]
@LaurentGoderre I agree with you. At that case, may be we add on the lighbox WET 4 example page a warning to avoid using lighbox with forms that require use input?

**Comment by pjackson28** [comment]
I don't think we need to be as drastic as telling users that they shouldn't use forms in the Lightbox. The main problem here is that Jaws is suppressing DOM focus events and most keyboard events (except with tabbing I think) so the scripting has no way to detect that the user has left the Lightbox (so doesn't know to close it). This is not a problem in NVDA. We do have an end of Lightbox link already but there is no way to cover all the possible ways of leaving the Lightbox (particularly with shortcuts).

The only known workaround for Jaws is to apply aria-hidden="true" to all ancestor elements outside of the Lightbox (to make them undiscoverable) but as you can imagine that could be quite risky (and is a very brute force approach).

Ultimately this is something that should be fixed at least in part by Jaws. Suppressing so many events is making it very difficult to properly support screen reader users.

**Comment by hsrudnicki** [comment]
Can this issue be closed?

**Comment by RubyDo** [comment]
Hello,  I would like to open the issue again. There is a focus issue for the WET 4 Lightbox. 
Demo showing the issue: https://youtu.be/pfUdagkL3JQ
Scenario: In lightbox, screen reader user accessing the search field using shortcut. 
Notice:
- Jaws indicates focus move to edit field 
- User could not complete search 
- After skipping the lightbox using esc key, the background shows focus is still on the link (that opened the lightbox) and not on the edit field

Note: @pjackson28 I believe the issue is not related to Jaws but more to the WEt 4 lightbox. 
I tried the same scenario on different lightbox (e.g. JIRA software - create issue lightbox) and it worked well ; the focus moved to the search edit field when selected with Jaws shortcut

**Comment by pjackson28** [comment]
@RubyDo As described above, JAWS is in part triggering this issue by suppressing DOM focus and keyboard events so WET currently is unable to tell when the focus is outside of the lightbox. This is not the case in NVDA as it doesn't suppress DOM focus events like JAWS does. The issue with the search box would go a way if there was some way in JavaScript to tell when the JAWS user has left the lightbox (as the lightbox remaining open is interfering).

Now perhaps JIRA is less interfering when the focus is outside of the dialog, but the problem still exists that the user is interacting with other parts of the page while the dialog is still open which could lead to any number of bugs in an application (since the user could change the state of something outside of the lightbox that is linked to what is being handled within the lightbox). So making the WET lightbox more friendly to navigation outside of the lightbox (while the lightbox is open) could be a bandaid for the reported symptom, but it doesn't address the underlying issue and could lead to much worse problems.

So I suggest that the focus should be on either preventing the user from leaving the lightbox while it remains open (not ideal) or finding a way to close the lightbox once the user has left regardless of how the lightbox is exited (ideal and what is done when JAWS is not involved). The problem though is that nobody has found a way to implement either of these approaches that covers all the way the user can leave the lightbox in JAWS (since there seems to be no way through JavaScript to tell when the JAWS user is outside of the lightbox). Hopefully someone can find a solution so the lightbox can be closed for JAWS once the user exits the lightbox (like for NVDA and other scenarios where JAWS isn't involved) as that would resolve the issue you reported as well without introducing other issues.

**Comment by hsrudnicki** [comment]
Based on the above, as well the milestone set, can be this issue closed as resolved/fixed or otherwise?

**Comment by RubyDo** [comment]
The issue is still existing but closed until someone finds a way to fix it or implement  approaches suggested above.

**Comment by RubyDo** [comment]
I reopened the issue. @pjackson28 Solutions to keep the keyboard focus on the lighbox and prevent users from accessing the keyboard were presented in a11yQC 2016 and a11yTO 2016. 
Here are links to the solutions: 
- a11yTO, Nov 5, 2016 by Bobby Bristol  https://twitter.com/RubysDo/status/794938054809157632
- a11yQC, Oct 11, 2016 by Jean-Marie D’amour & Guillaume D’amour http://labo.raamm.org/test/test-4-boites-de-dialogue/ 
I assume this fix would be required for the WET4 Session Timeout as well

**Comment by pjackson28** [comment]
@RubyDo Thank you for the update. The <code>aria-hidden="true"</code> was raised previously as a possible workaround (see my March 2015 comment), but it may be tricky depending upon where the lightbox is within the DOM in relation to the elements that will be affected by <code>aria-hidden="true"</code>. For instance if the lightbox is a descendant of the node that <code>aria-hidden="true"</code> is applied to then there is the possibility that it could impact the accessibility of the lightbox (would require testing across different screen readers to confirm whether or not that is the case). The ideal scenario would be to ensure that the lightbox is not a descendent of the nodes that <code>aria-hidden="true"</code> is being applied to, like I've seen in some examples in the past, but I don't know how achievable that is in WET.

The <code>aria-hidden</code> may end up being the best solution but someone would need to code and test this approach to confirm for sure. Do you know anyone that is willing to take on this work? I don't lead the WET project anymore so I can't speak for any of the resources on the WET project team.

**Comment by RubyDo** [comment]
@pjackson28 Thanks for the clarification and the directions on how to fix it. 
We may have a new member in our team in January available to contribute on fixing the lightbox issue on WET.

**Comment by RubyDo** [comment]
Hello @pjackson28 @masterbee , we use the lighbox very often and are willing to contribute in the fixing of the lighbox issue. @kevinLamb joined us our team this winter and he is looking into fixing it.

**Comment by duboisp** [comment]
@RubyDo Can you re-test to see if it still an issue. Thanks

**Comment by RubyDo** [comment]
@duboisp  I tested the lighbox on WET 4 v25 for an app we are developing and it works fine on JAWS 15 and above. However, my colleague Daniel found an issue on NVDA.

Scenario as described by Dan on NVDA: User opens a lighbox with close button. User opens list of buttons using the quick navigation key, User can navigate to close button as well as other buttons on the background page but focus is quickly brought back to the “X” button within the lightbox.
